### ISSUE TITLE
[BUG] Wrong limb width when cross compiling

### ISSUE THREAD

**Issue by dtolnay** [Alexhuszagh/rust-lexical#43]
The following code in the lexical-core build script branches on the value of cfg(target_arch). That cfg refers to the target of the current code being compiled by rustc. In the case of the build script that's the host architecture.

https://github.com/Alexhuszagh/rust-lexical/blob/8c75e297689591fea903176bcaeb03c27325dccf/lexical-core/build.rs#L10-L20

Cargo provides a separate env variable to build scripts, called CARGO_CFG_TARGET_ARCH, to determine the target arch of the library as opposed to the target arch of the build script.

To reproduce, put this in build.rs and run `cargo check --target wasm32-unknown-unknown` and see which error is triggered.

```rust
#[cfg(target_arch = "x86_64")]
compile_error!("target_arch = x86_64");
#[cfg(target_arch = "wasm32")]
compile_error!("target_arch = wasm32");
```

**Comment by myrrlyn** [comment]
Is this the correct resolution?

**Comment by dtolnay** [comment]
No. The target arch **of the build script** should have no bearing. Only the target arch which the **library** is being built for is relevant.

**Comment by dtolnay** [comment]
In other words absence of `CARGO_CFG_TARGET_ARCH` needs to be either a build error (preferred), or should guess a consistent limb width regardless of what the host architecture happens to be.

**Comment by myrrlyn** [comment]
I was about to ask what the resolution should be in the `Err` case. I'll fail the build.

**Comment by myrrlyn** [comment]
Once you're satisfied that this is correct, I'll publish an update, probably in May.

I don't know the versioning history of `lexical-core` or what its current dependents are. Do you know of a need for me to backport to `0.6`, or will an `0.7` patch release suffice?

**Comment by dtolnay** [comment]
Thanks -- Personally I don't use lexical-core so I don't have any stake in a release. I fixed this a long time ago on my end in https://github.com/serde-rs/json/commit/6e00729f1f610eac1087678d80613f6979da08ba. Serde uses a fork because I wasn't confident in the maintenance of this crate.

**Comment by Alexhuszagh** [comment]
@dtolnay I'm glad it was of use to you, sorry, for mental health reasons I haven't exactly been the best maintainer so I'm very glad an adaptation has suited your needs. Thank you for reporting this. 

Are there any cases where `CARGO_CFG_TARGET_ARCH` will not be defined? If so, it's pretty easy to default to 32-bit limbs if the environment variable is not set. We use an optimization when 64-bit integer operations are faster on the CPU, but if that's unknown, it would be better to default to 32-bit limbs than error during build.

**Comment by Alexhuszagh** [comment]
I think I've satisfactorily fixed this issue. I've removed the `target_arch` of the build script from `build.rs`, and defaulted to 32-bit limbs if `CARGO_CFG_TARGET_ARCH ` is not set, since that should be reasonably performant on most architectures.

**Comment by Alexhuszagh** [comment]
Closed as of [v0.4.8](https://github.com/Alexhuszagh/rust-lexical/releases/tag/v0.4.8), [v0.5.1](https://github.com/Alexhuszagh/rust-lexical/releases/tag/v0.5.1), [v0.6.8](https://github.com/Alexhuszagh/rust-lexical/releases/tag/v0.6.8), and [v0.7.6](https://github.com/Alexhuszagh/rust-lexical/releases/tag/v0.7.6).
